src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,150 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImport;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Packaging;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.SymbolSearch;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.AddMissingImports
+{
+    internal abstract class AbstractAddMissingImportsFeatureService : IAddMissingImportsFeatureService
+    {
+        protected abstract ImmutableArray<string> FixableDiagnosticIds { get; }
+
+        public async Task<Project> AddMissingImportsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            // Get the diagnostics that indicate a missing import.
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (diagnostics.IsEmpty)
+            {
+                return document.Project;
+            }
+
+            // Find fixes for the diagnostic where there is only a single fix.
+            var usableFixes = await GetUnambiguousFixesAsync(document, diagnostics, cancellationToken).ConfigureAwait(false);
+            if (usableFixes.IsEmpty)
+            {
+                return document.Project;
+            }
+
+            // Apply those fixes to the document.
+            var newDocument = await ApplyFixesAsync(document, usableFixes, cancellationToken).ConfigureAwait(false);
+            return newDocument.Project;
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var project = document.Project;
+            var compilation = await project.GetCompilationAsync(cancellationToken);
+
+            // Ensure that diagnostics for this document are always in document location
+            // order.  This provides a consistent and deterministic order for fixers
+            // that want to update a document.
+            var diagnostics = compilation.GetDiagnostics(cancellationToken)
```

Comment:
wow.  that's super expensive.  this is *all* diagnostics for the compilation.  Why not get this for the semantic model for the document you're on?   SemanticModel.GetDiagnostics also takes a textSpan, so it will do this efficiently.